src/main/java/com/thealgorithms/datastructures/buffers/DoubleBuffer.java

```diff
@@ -0,0 +1,105 @@
+package com.thealgorithms.datastructures.buffers;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * The {@code DoubleBuffer} class implements a double buffering mechanism.
+ * It maintains two buffers, allowing the program to switch between them.
+ * This class is useful for separating read/write operations and optimizing memory usage.
+ *
+ * @param <Item> The type of elements stored in the double buffer.
+ */
+public class DoubleBuffer<Item> {
+    private final Item[] primaryBuffer;
+    private final Item[] secondaryBuffer;
+    private final AtomicInteger primaryIndex;
+    private final AtomicInteger secondaryIndex;
+    private final int capacity;
+    private final AtomicBoolean isPrimaryActive;
+
+    /**
+     * Constructor to initialize the double buffer with a specified capacity.
+     *
+     * @param capacity The size of each buffer.
+     * @throws IllegalArgumentException if the capacity is zero or negative.
+     */
+    public DoubleBuffer(int capacity) {
+        if (capacity <= 0) {
+            throw new IllegalArgumentException("Buffer size must be positive");
+        }
+        // noinspection unchecked
+        this.primaryBuffer = (Item[]) new Object[capacity];
+        this.secondaryBuffer = (Item[]) new Object[capacity];
+        this.primaryIndex = new AtomicInteger(0);
+        this.secondaryIndex = new AtomicInteger(0);
+        this.capacity = capacity;
+        this.isPrimaryActive = new AtomicBoolean(true);
+    }
+
+    /**
+     * Checks if the active buffer is empty.
+     *
+     * @return {@code true} if the active buffer is empty, {@code false} otherwise.
+     */
+    public boolean isEmpty() {
+        return isPrimaryActive.get() ? primaryIndex.get() == 0 : secondaryIndex.get() == 0;
+    }
+
+    /**
+     * Switches between the primary and secondary buffers.
+     */
+    public void switchBuffer() {
+        isPrimaryActive.set(!isPrimaryActive.get());
+    }
+
+    /**
+     * Checks if the primary buffer is currently active.
+     *
+     * @return {@code true} if the primary buffer is active, {@code false} otherwise.
+     */
+    public boolean isPrimaryActive() {
+        return isPrimaryActive.get();
+    }
+
+    /**
+     * Adds an item to the active buffer.
+     *
+     * @param item The item to be added.
+     * @throws IllegalArgumentException if the item is null.
+     */
+    public void put(Item item) {
+        if (item == null) {
+            throw new IllegalArgumentException("Null items are not allowed");
+        }
+
+        if (isPrimaryActive.get()) {
+            if (primaryIndex.get() < capacity) {
+                primaryBuffer[primaryIndex.getAndIncrement()] = item;
```

Comment:
As you want to make this code thread-safe, will be better to use `ReentrantLock` for this.


```
private final ReentrantLock lock = new ReentrantLock();

public void switchBuffer() {
    lock.lock();
    try {
        isPrimaryActive.set(!isPrimaryActive.get());
    } finally {
        lock.unlock();
    }
}

public boolean put(Item item) {
    if (item == null) {
        throw new IllegalArgumentException("Null items are not allowed");
    }
    lock.lock();
    try {
        if (isPrimaryActive.get()) {
            if (primaryIndex.get() < capacity) {
                primaryBuffer[primaryIndex.getAndIncrement()] = item;
                return true;
            }
        } else {
            if (secondaryIndex.get() < capacity) {
                secondaryBuffer[secondaryIndex.getAndIncrement()] = item;
                return true;
            }
        }
        return false; // Buffer is full
    } finally {
        lock.unlock();
    }
}

public Item get() {
    lock.lock();
    try {
        if (isPrimaryActive.get()) {
            if (primaryIndex.get() == 0) {
                return null;
            }
            return primaryBuffer[primaryIndex.decrementAndGet()];
        } else {
            if (secondaryIndex.get() == 0) {
                return null;
            }
            return secondaryBuffer[secondaryIndex.decrementAndGet()];
        }
    } finally {
        lock.unlock();
    }
}
```